File: src/main/store/jsonStore.ts
import { readFileSync, writeFileSync, renameSync, mkdirSync, copyFileSync } from 'fs'
import { dirname } from 'path'
import { createLogger } from '../core/logger'

const log = createLogger('store')

// ── Dayanıklı JSON deposu ──────────────────────────────────────────────────
// Tüm kalıcı depolar (sites, known_hosts, trusted_certs) için ortak kurallar:
//  1. Atomik yazma: önce .tmp'ye yaz, sonra rename — yarım yazım ana dosyayı bozamaz.
//  2. Sürümlü zarf: { version, data } — ileride şema değişince migrate edilebilir.
//  3. Bozulma karantinası: parse edilemeyen dosya .corrupt olarak SAKLANIR
//     (üzerine yazılıp veri kaybedilmez), depo boş/varsayılan olarak açılır.

interface Envelope {
  version: number
  data: unknown
}

function isEnvelope(v: unknown): v is Envelope {
  return (
    typeof v === 'object' &&
    v !== null &&
    !Array.isArray(v) &&
    typeof (v as Envelope).version === 'number' &&
    'data' in v
  )
}

/** Değeri {version, data} zarfıyla atomik olarak (tmp + rename) yazar. */
export function writeJsonVersioned(path: string, version: number, data: unknown): void {
  mkdirSync(dirname(path), { recursive: true })
  const tmp = `${path}.tmp`
  writeFileSync(tmp, JSON.stringify({ version, data } satisfies Envelope, null, 2), 'utf8')
  renameSync(tmp, path)
}

export interface ReadOptions<T> {
  /**
   * Zarf öncesi eski (çıplak) formatı güncel şemaya taşır. Tanıyamazsa null
   * döndürmelidir; null → dosya bozuk kabul edilir ve karantinaya alınır.
   */
  legacy?: (parsed: unknown) => T | null
  /** Eski zarf sürümünü güncel şemaya taşır. Tanıyamazsa null döndürmelidir. */
  migrate?: (fileVersion: number, data: unknown) => T | null
  /**
   * Bozuk dosya karantinaya alındıktan sonra dönecek değer. Verilmezse
   * onMissing() kullanılır. "İlk çalıştırma" ile "bozulma"nın farklı davranması
   * gerektiğinde (örn. bozulmada seed İÇE AKTARMAMAK için) kullanılır.
   */
  onCorrupt?: () => T
}

/**
 * Sürümlü depoyu okur.
 *  - Dosya yoksa: onMissing() (ilk çalıştırma).
 *  - Güncel sürümse: data.
 *  - Eski zarf sürümü / eski çıplak format: migrate()/legacy() ile taşınır.
 *  - Bozuksa: dosya `.corrupt` kopyasına alınır ve onMissing() döner — böylece
 *    bozulan veri bir sonraki save()'de sessizce silinmez, kurtarılabilir kalır.
 */
export function readJsonVersioned<T>(
  path: string,
  version: number,
  onMissing: () => T,
  opts: ReadOptions<T> = {}
): T {
  let raw: string
  try {
    raw = readFileSync(path, 'utf8')
  } catch {
    return onMissing() // dosya yok → ilk çalıştırma
  }

  try {
    const parsed = JSON.parse(raw) as unknown
    if (isEnvelope(parsed)) {
      if (parsed.version === version) return parsed.data as T
      const migrated = opts.migrate?.(parsed.version, parsed.data) ?? null
      if (migrated !== null) {
        log.info(`${path}: v${parsed.version} → v${version} taşındı`)
        return migrated
      }
      throw new Error(`desteklenmeyen depo sürümü: ${parsed.version}`)
    }
    const legacy = opts.legacy?.(parsed) ?? null
    if (legacy !== null) {
      log.info(`${path}: eski format güncel şemaya taşındı`)
      return legacy
    }
    throw new Error('tanınmayan depo biçimi')
  } catch (err) {
    const quarantine = `${path}.corrupt`
    try {
      copyFileSync(path, quarantine)
      log.error(`${path} bozuk — kopya ${quarantine} olarak saklandı`, String(err))
    } catch (copyErr) {
      log.error(`${path} bozuk ve karantinaya alınamadı`, String(copyErr))
    }
    return (opts.onCorrupt ?? onMissing)()
  }
}

File: src-tauri/src/store/json_store.rs
//! Dayanıklı JSON deposu — `src/main/store/jsonStore.ts` karşılığı.
//!  1. Atomik yazma: önce `.tmp`'ye yaz, sonra rename.
//!  2. Sürümlü zarf: `{ version, data }`.
//!  3. Bozulma karantinası: parse edilemeyen dosya `.corrupt` olarak SAKLANIR,
//!     depo varsayılan/boş açılır (kullanıcı verisi silinmez).

use std::fs;
use std::path::Path;

use serde::de::DeserializeOwned;
use serde::Serialize;
use serde_json::Value;

/// Değeri `{version, data}` zarfıyla atomik (tmp + rename) yazar.
pub fn write_versioned<T: Serialize>(
    path: &Path,
    version: u32,
    data: &T,
) -> std::io::Result<()> {
    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent)?;
    }
    let envelope = serde_json::json!({ "version": version, "data": data });
    let tmp = path.with_extension("json.tmp");
    let text = serde_json::to_string_pretty(&envelope)
        .map_err(|e| std::io::Error::new(std::io::ErrorKind::InvalidData, e))?;
    fs::write(&tmp, text)?;
    fs::rename(&tmp, path)?;
    Ok(())
}

/// Sürümlü/atomik okuma sonucu.
pub enum ReadOutcome<T> {
    /// Dosya yok — ilk çalıştırma.
    Missing,
    /// Güncel sürüm başarıyla okundu.
    Loaded(T),
    /// Dosya vardı ama bozuktu; `.corrupt`'a alındı.
    Corrupt,
}

/// Sürümlü depoyu okur. `legacy` verilirse zarf öncesi (çıplak) format da denenir.
///
/// - Dosya yok → `Missing`.
/// - Güncel sürüm zarfı → `Loaded`.
/// - Zarf ama farklı sürüm ya da çıplak legacy format → `legacy(parsed)` denenir.
/// - Aksi halde `.corrupt`'a kopyalanır → `Corrupt`.
pub fn read_versioned<T, F>(path: &Path, version: u32, legacy: F) -> ReadOutcome<T>
where
    T: DeserializeOwned,
    F: FnOnce(&Value) -> Option<T>,
{
    let raw = match fs::read_to_string(path) {
        Ok(s) => s,
        Err(_) => return ReadOutcome::Missing,
    };

    let parsed: Value = match serde_json::from_str(&raw) {
        Ok(v) => v,
        Err(_) => return quarantine(path),
    };

    // Zarf mı? { version: number, data: ... }
    if let Some(obj) = parsed.as_object() {
        if let (Some(v), true) = (obj.get("version").and_then(Value::as_u64), obj.contains_key("data")) {
            if v as u32 == version {
                return match serde_json::from_value::<T>(obj["data"].clone()) {
                    Ok(data) => ReadOutcome::Loaded(data),
                    Err(_) => quarantine(path),
                };
            }
            // Farklı sürüm — legacy migrasyonu dene, olmazsa karantina.
            return match legacy(&parsed) {
                Some(data) => ReadOutcome::Loaded(data),
                None => quarantine(path),
            };
        }
    }

    // Çıplak (zarfsız) eski format.
    match legacy(&parsed) {
        Some(data) => ReadOutcome::Loaded(data),
        None => quarantine(path),
    }
}

/// Bozuk dosyayı `.corrupt` kopyasına alır ve `Corrupt` döner.
fn quarantine<T>(path: &Path) -> ReadOutcome<T> {
    let dst = path.with_extension("json.corrupt");
    if let Err(e) = fs::copy(path, &dst) {
        eprintln!("[ferro] {path:?} bozuk ve karantinaya alınamadı: {e}");
    } else {
        eprintln!("[ferro] {path:?} bozuk — kopya {dst:?} olarak saklandı");
    }
    ReadOutcome::Corrupt
}
